schema/query/projection_evaluator.go

```diff
@@ -4,24 +4,27 @@ import (
 	"context"
 	"fmt"
 
+	"sync"
+
 	"github.com/rs/rest-layer/schema"
 )
 
 // Eval evaluate the projection on the given payload with the help of the
 // validator. The resolver is used to fetch payload of references outside of the
 // provided payload.
 func (p Projection) Eval(ctx context.Context, payload map[string]interface{}, validator schema.Validator, resolver ReferenceResolver) (map[string]interface{}, error) {
-	payload, err := evalProjection(ctx, p, payload, validator, resolver)
+	rbr := &referenceBatchResolver{}
+	payload, err := evalProjection(ctx, p, payload, validator, rbr)
 	if err == nil {
-		// The resulting payload may contain some asyncProjection, we must execute
-		// them concurrently until there's no more.
-		err = resolveAsyncProjections(ctx, payload, resolver)
+		// Execute the batched reference resolutions.
+		err = rbr.execute(ctx, resolver)
 	}
 	return payload, err
 }
 
-func evalProjection(ctx context.Context, p Projection, payload map[string]interface{}, validator schema.Validator, resolver ReferenceResolver) (map[string]interface{}, error) {
+func evalProjection(ctx context.Context, p Projection, payload map[string]interface{}, validator schema.Validator, rbr *referenceBatchResolver) (map[string]interface{}, error) {
 	res := map[string]interface{}{}
+	resMu := sync.Mutex{} // XXX use sync.Map once Go 1.9 is out
```

Comment:
Maybe.  Although `sync.Map` is very cool, it's a very complex implementation (from what little I have seen so far`), that has some extra allocates, e.g. to achieve lock-free reads.

Then functions under evalProjection seams to be mostly _writing_ to the map. If the mutex can be used right, it's not _necessarily_ true that a `sync.Map` will be faster, although it's somewhat easier to not screw it up.  It might be good with a benchmark to help make such a decision.